src/sunmatrix/dense/sunmatrix_dense.c

```diff
@@ -325,11 +325,45 @@ SUNErrCode SUNMatMatvec_Dense(SUNMatrix A, N_Vector x, N_Vector y)
   SUNAssert(xd != yd, SUN_ERR_MEM_FAIL);
 
   /* Perform operation y = Ax */
-  for (i = 0; i < SM_ROWS_D(A); i++) { yd[i] = ZERO; }
-  for (j = 0; j < SM_COLUMNS_D(A); j++)
+  for (sunindextype i = 0; i < SM_ROWS_D(A); i++) { yd[i] = ZERO; }
+  for (sunindextype j = 0; j < SM_COLUMNS_D(A); j++)
+  {
+    sunrealtype* col_j = SM_COLUMN_D(A, j);
+    for (sunindextype i = 0; i < SM_ROWS_D(A); i++)
+    {
+      yd[i] += col_j[i] * xd[j];
+    }
+  }
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNMatMatvecTranspose_Dense(SUNMatrix A, N_Vector x, N_Vector y)
+{
+  SUNFunctionBegin(A->sunctx);
+
+  SUNAssert(SUNMatGetID(A) == SUNMATRIX_DENSE, SUN_ERR_ARG_WRONGTYPE);
+  SUNCheck(compatibleMatrixAndVectors(A, y, x), SUN_ERR_ARG_DIMSMISMATCH);
+
+  /* access vector data (return if NULL data pointers) */
+  sunrealtype *xd, *yd;
+  xd = N_VGetArrayPointer(x);
+  SUNCheckLastErr();
+  yd = N_VGetArrayPointer(y);
+  SUNCheckLastErr();
+
+  SUNAssert(xd, SUN_ERR_MEM_FAIL);
+  SUNAssert(yd, SUN_ERR_MEM_FAIL);
+  SUNAssert(xd != yd, SUN_ERR_MEM_FAIL);
+
+  /* Perform operation y = A^T x */
+  for (sunindextype i = 0; i < SM_COLUMNS_D(A); i++) { yd[i] = ZERO; }
+  for (sunindextype i = 0; i < SM_COLUMNS_D(A); i++)
   {
-    col_j = SM_COLUMN_D(A, j);
-    for (i = 0; i < SM_ROWS_D(A); i++) { yd[i] += col_j[i] * xd[j]; }
+    sunrealtype* row_i = SM_COLUMN_D(A, i);
+    for (sunindextype j = 0; j < SM_ROWS_D(A); j++)
```

Comment:
This looks fine for square matrices, but for a rectangular matrix shouldn't the loop upper bounds be swapped?
```
  for (sunindextype i = 0; i < SM_ROWS_D(A); i++)
  {
    sunrealtype* row_i = SM_COLUMN_D(A, i);
    for (sunindextype j = 0; j < SM_COLUMNS_D(A); j++)
```

Reply:
I am not seeing why the bounds would need to be swapped for a rectangular matrix, and the code currently passes the tests with both square and rectangular matrices.